cpp/src/arrow/compute/kernels/aggregate_test.cc

```diff
@@ -873,6 +874,98 @@ TYPED_TEST(TestRandomNumericCountKernel, RandomArrayCount) {
   }
 }
 
+//
+// Count Distinct
+//
+
+class TestCountDistinctKernel : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    only_valid = CountOptions(CountOptions::ONLY_VALID);
+    only_null = CountOptions(CountOptions::ONLY_NULL);
+    all = CountOptions(CountOptions::ALL);
+  }
+
+  Datum Expected(int64_t value) { return MakeScalar(static_cast<int64_t>(value)); }
+
+  void Check(Datum input, int64_t expected_all, bool has_nulls = true) {
+    int64_t expected_valid = has_nulls ? expected_all - 1 : expected_all;
+    int64_t expected_null = has_nulls ? 1 : 0;
+    CheckScalar("count_distinct", {input}, Expected(expected_valid), &only_valid);
+    CheckScalar("count_distinct", {input}, Expected(expected_null), &only_null);
+    CheckScalar("count_distinct", {input}, Expected(expected_all), &all);
+  }
+
+  void Check(const std::shared_ptr<DataType>& type, util::string_view json,
+             int64_t expected_all, bool has_nulls = true) {
+    Check(ArrayFromJSON(type, json), expected_all, has_nulls);
+  }
+
+  CountOptions only_valid;
+  CountOptions only_null;
+  CountOptions all;
+};
+
+TEST_F(TestCountDistinctKernel, AllArrowTypesWithNulls) {
```

Comment:
Can you add a test for scalars?

Reply:
I'm using CheckScalar for this tests, my understating was that this will cover the scalar tests too.
https://github.com/apache/arrow/pull/11159#discussion_r710215610

> Use CheckScalar. As noted below, you'll need to add support for all-scalar inputs. The advantage is that CheckScalar automatically also tests chunked array and scalar inputs for you.

Reply:
Hmm, I see. That comment only applies when the tested function is a scalar function (with array output). Here you are testing an aggregate function (with scalar output), so `CheckScalar` has no way of knowing how to compute the expected output for subsets of the input.

Actually, we should probably guard against uses of `CheckScalar` with a non-scalar function, since that's an incorrect use of this API.

Reply:
As a side note, if there's an apparent bug in the implementation, and the bug isn't surfaced in the tests, then logically the tests probably don't cover that bug.

Reply:
Ah, that's my bad, sorry for the confusion.

Reply:
Well, there are three different kinds of "count_distinct" (count all, count nulls, count non-nulls), so that makes the output slightly less trivial :-)

(also, ideally, we would also test with a non-trivial `batch.length`)

Reply:
> (also, ideally, we would also test with a non-trivial batch.length)

Yes, that is the test I want to write for this PR but it seems there isn't any test util to make that simply.

Reply:
See https://github.com/apache/arrow/blob/dfb109c0a037a4f9bd5df6e99ce6425e2cb618c3/cpp/src/arrow/compute/exec/plan_test.cc#L791

Indeed it's not easy to test since ExecBatch isn't in the public API. Effectively it's only possible through an ExecPlan.

Reply:
In that case we should at least test the more trivial cases. Ping @lidavidm on testing `batch.length`.

Reply:
> Well, there are three different kinds of "count_distinct" (count all, count nulls, count non-nulls), so that makes the output slightly less trivial :-)

Still I don't follow what specific test do I need to write here, do you think I will need to write scalar tests in this PR for all those kernels? If so, what is the template I can follow for that?

Reply:
Not sure whether you saw my reply in the other thread: https://github.com/apache/arrow/pull/11257#discussion_r721484260

Reply:
There are tests for count: https://github.com/apache/arrow/blob/dfb109c0a037a4f9bd5df6e99ce6425e2cb618c3/cpp/src/arrow/compute/kernels/aggregate_test.cc#L846-L856

Reply:
Got it! so something like this should pass right?
```cpp
EXPECT_THAT(CallFunction("count_distinct", {MakeScalar(true)}, &all), ResultWith(Datum(std::make_shared<UInt64Scalar>(1))));
```

Reply:
Failed, with this message. I think I'll need to spend some time to see what is the real issue here (it would help to have better messages for the output :))
```cpp
Value of: CallFunction("count_distinct", {MakeScalar(true)}, &all)
Expected: value is equal to 1
  Actual: 1, whose value 1 doesn't match
[  FAILED  ] TestCountDistinctKernel.ScalarTypesWithNulls (0 ms)
```

Reply:
Ah, that's a bit unfortunate. I guess the output type is not the same (UInt64 vs. Int64 presumably?).

Reply:
Yes, I'm getting used to those error messages, it was the output type, thanks!

Reply:
Done, thanks!